File: best_sum.py
#!/usr/bin/env python3

all_combinations = []
best_combination = []

def best_sum(target, nums, lt):
    if target == 0:
        all_combinations.append(lt.copy())
        if len(lt) > len(best_combination):
            best_combination = lt.copy()
        return
    if target < 0:
        # lt.pop()
        return

    print('---', target, nums, lt)

    for num in nums:
        rem = target - num
        lt.append(num)
        print(rem, lt, num)
        x = best_sum(rem, nums, lt)
        lt.pop()


op = best_sum(7, [5, 3, 4, 7], [])
print(all_combinations)
print(best_combination)

File: best_sum.go
package main

import (
	"fmt"
)

var best_sum []int

func get_best_sum(target int, nums []int, lt []int) {
	// fmt.Print(target, nums, lt)
	if target == 0 {
		if best_sum == nil || len(best_sum) > len(lt) {
			best_sum = append([]int{}, lt...)
		}
		return
	}
	if target < 0 {
		return
	}
	for _, num := range nums {
		rem := target - num
		lt = append(lt, num)
		get_best_sum(rem, nums, lt)
		lt = lt[:len(lt)-1]
	}

}

// func main() {
// 	get_best_sum(7, []int{5, 3, 4, 7}, []int{})
// 	fmt.Print(best_sum)
// }
